compute item to get L2-norm of tensors

prayush opened the pull request:
## Proposed changes

Added a compute item to get L2-norm of tensors. This would be useful for observing constraints



### Types of changes:

- [x] Bugfix
- [x] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

prayush committed:
Compute item to get L2-norm of tensors

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -48,11 +50,38 @@ Scalar<DataType> magnitude(
+template <typename DataType, typename Symm, typename IndexList>
+Scalar<DataType> l2_norm(
+    const Tensor<DataType, Symm, IndexList>& tensor) noexcept {
+  auto& input = const_cast<Tensor<DataType, Symm, IndexList>&>(tensor);
```
Can you remove this line?

prayush replied:
removed

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -48,11 +50,38 @@ Scalar<DataType> magnitude(
+Scalar<DataType> l2_norm(
+    const Tensor<DataType, Symm, IndexList>& tensor) noexcept {
+  auto& input = const_cast<Tensor<DataType, Symm, IndexList>&>(tensor);
+  DataType sum(tensor.begin()->size(), 0.);
```
Prefer `{}` initialization. Also, name more descriptively, e.g. `pointwise_l2_norm` and/or add a comment what this is.

prayush replied:
changed

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -48,11 +50,38 @@ Scalar<DataType> magnitude(
+    const Tensor<DataType, Symm, IndexList>& tensor) noexcept {
+  auto& input = const_cast<Tensor<DataType, Symm, IndexList>&>(tensor);
+  DataType sum(tensor.begin()->size(), 0.);
+  for (auto it = input.begin(); it != input.end(); ++it) {
```
Can you make this a `for (auto& tensor_element : tensor)` loop?

prayush replied:
done

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -48,11 +50,38 @@ Scalar<DataType> magnitude(
+  auto& input = const_cast<Tensor<DataType, Symm, IndexList>&>(tensor);
+  DataType sum(tensor.begin()->size(), 0.);
+  for (auto it = input.begin(); it != input.end(); ++it) {
+    sum += input.multiplicity(it) * square(*it);
```
Do you do this to handle tensor symmetries? Could you add a comment?

prayush replied:
yep, comment added.

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -48,11 +50,38 @@ Scalar<DataType> magnitude(
+  for (auto it = input.begin(); it != input.end(); ++it) {
+    sum += input.multiplicity(it) * square(*it);
+  }
+  return Scalar<DataType>{sqrt(sum)};
```
Why do you have this `Scalar` overload of the function and not only the `double`, what would you want to use the pointwise L2-norm for?

prayush replied:
I want to be able to visualize pointwise L2 norms of constraints over the domain. this helps spatially localize constraint violations and understand their kinematics.

nilsvu replied:
Ok sounds useful :) thanks.

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -141,4 +170,17 @@ struct Sqrt : db::ComputeTag {
+///
+/// \snippet Test_Magnitude.cpp sqrt_name
+template <typename Tag, bool Accumulate = false>
+struct L2Norm : db::ComputeTag {
```
Perhaps we should have a compute item or at least a free function for the square of the L2 norm so that we can use it in `ObserveErrorNorms`. @wthrowe what do you think?

prayush replied:
Do we really want a free function to perform a `square` operation - isn't calling `square(db::get<L2Norm<Tag>>(box))` equivalent?

prayush replied:
An alternate could be to redefine these compute tags as `L2NormSq` and `PointwiseL2NormSq`, which can then be clubbed with the compute tag `Sqrt` to do what i want, or not to use it in `ObserveErrornorms`. Although, i'm not sure i like this idea much, but still .. thoughts?

prayush replied:
Just added a function for the square of the point-wise L2 norm.

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -141,4 +170,17 @@ struct Sqrt : db::ComputeTag {
+/// \snippet Test_Magnitude.cpp sqrt_name
+template <typename Tag, bool Accumulate = false>
+struct L2Norm : db::ComputeTag {
+  static std::string name() noexcept { return "L2Norm(" + Tag::name() + ")"; }
```
`db::tag_name<Tag>()`

prayush replied:
changed

nilsvu commented on tests/Unit/DataStructures/Tensor/EagerMath/Test_Magnitude.cpp:
```diff
@@ -261,6 +267,82 @@ void test_root_tags() {
+
+void test_l2_norm_tag() {
+  constexpr size_t npts = 5;
+  MAKE_GENERATOR(generator);
```
I don't think you should do a random-value test that re-implements parts of the code you want to test. I'd prefer to start with hard-coded numbers and compare to other hard-coded numbers.

prayush replied:
I am a bit confused by what you mean when you say that the test _re-implements parts of the code you want to test_. 

The code to be tested computes the norm by looping over (only) those tensor components that are stored in memory, and uses their multiplicity; it never explicitly indexes the tensor whose norm is being taken. The test, on the other hand, accesses tensor components through explicit indexing. There is no loop over components in the test either. Which part of the tested code is being "re-implemented" in the test? The `square` and `add` operations?

nilsvu replied:
It re-implements the `PlusSquare` and `alg::accumulate`.

Generally, I'm worried when I see a test with no hard-coded values. And I don't see the benefit of random-value tests here since this implements no particularly complicated set of equations where random numbers could turn up edge-cases in the numerical implementation.

prayush replied:
I see what you mean, and have replaced all randomized tests with tests using hard-coded numbers for input/output.

fmahebert reviewed (commented):
After having computed the pointwise norm "x" from the various components,
- SpEC's L2 norm is `sqrt{ sum_gridpoints{ x^2 } / number_of_gridpoints }`
- this (AFAICT) computes `sum_gridpoints{ x^2 }`

Am I misreading the code? Or is there a particular reason to change how we compute norms?

I would also suggest separating the computation of the pointwise norm from the computation of the norm over gridpoints.

prayush commented:
> Am I misreading the code? Or is there a particular reason to change how we compute norms?
> 
There should have been a `sqrt( ... / num_of_gridpoints)` operation which I've now added. I was using these compute tags to observe GH constraint norms' evolution, and I was doing this operation outside the compute tag there, so this bug crept in.

> I would also suggest separating the computation of the pointwise norm from the computation of the norm over gridpoints.

done. Now there are two compute tags, `PointwiseL2Norm` and `L2Norm`.

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -54,35 +54,51 @@ Scalar<DataType> sqrt_magnitude(const Scalar<DataType>& input) noexcept {
-  DataType sum(tensor.begin()->size(), 0.);
-  for (auto it = input.begin(); it != input.end(); ++it) {
-    sum += input.multiplicity(it) * square(*it);
+  DataType pointwise_l2_normsq{tensor.begin()->size(), 0.};
```
You can use `make_with_value` here

prayush replied:
changed

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -54,35 +54,51 @@ Scalar<DataType> sqrt_magnitude(const Scalar<DataType>& input) noexcept {
-  DataType sum(tensor.begin()->size(), 0.);
-  for (auto it = input.begin(); it != input.end(); ++it) {
-    sum += input.multiplicity(it) * square(*it);
+  DataType pointwise_l2_normsq{tensor.begin()->size(), 0.};
```
how about you move the shared code between these two functions into a `detail::pointwise_l2_norm_square` function?

prayush replied:
done

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -54,35 +54,51 @@ Scalar<DataType> sqrt_magnitude(const Scalar<DataType>& input) noexcept {
+    // pointer to non-const type.
   }
-  return Scalar<DataType>{sqrt(sum)};
+  return Scalar<DataType>{sqrt(pointwise_l2_normsq)};
```
Does one not usually divide by the number of tensor components here?

prayush replied:
not usually when computing the [Frobenius norm](https://en.wikipedia.org/wiki/Matrix_norm#Frobenius_norm).

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -175,14 +191,30 @@ struct Sqrt : db::ComputeTag {
+  static std::string name() noexcept {
+    return "PointwiseL2Norm(" + db::tag_name<Tag>() + ")";
+  }
+  static constexpr Scalar<DataVector> (*function)(const db::item_type<Tag>&) =
```
can you make this `auto` now that the functions have different names?

prayush replied:
Compiler doesn't have enough information to deduce the template parameter `DataType` automatically

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -48,11 +50,56 @@ Scalar<DataType> magnitude(
+        square(tensor_element);
+    // The `const_cast` above is necessary to stop gcc warning (error) about
+    // the implicit conversion of function argument from a pointer to const to
+    // pointer to non-const type.
```
Thanks for this comment, but I still don't quite understand why the `const_cast` is needed. The `multiplicity` function takes an iterator by const-reference. I don't know if what you are doing here and the `NOLINT` is the right thing to do. Perhaps someone from @sxs-collaboration/spectre-core-devs can have a quick look?

wthrowe replied:
You've found a bug in `Tensor`.  Apply this patch:
```patch
diff --git a/src/DataStructures/Tensor/Tensor.hpp b/src/DataStructures/Tensor/Tensor.hpp
index 31882e4f8..f5d3aa71e 100644
--- a/src/DataStructures/Tensor/Tensor.hpp
+++ b/src/DataStructures/Tensor/Tensor.hpp
@@ -353,7 +353,7 @@ class Tensor<X, Symm, IndexList<Indices...>> {
   ///
   /// \see TensorMetafunctions::compute_multiplicity
   SPECTRE_ALWAYS_INLINE constexpr size_t multiplicity(
-      const iterator& iter) const noexcept {
+      const const_iterator& iter) const noexcept {
     return structure::multiplicity(static_cast<size_t>(iter - begin()));
   }
   SPECTRE_ALWAYS_INLINE static constexpr size_t multiplicity(
diff --git a/tests/Unit/DataStructures/Tensor/Test_Tensor.cpp b/tests/Unit/DataStructures/Tensor/Test_Tensor.cpp
index bf89c22b4..2e2c2efe5 100644
--- a/tests/Unit/DataStructures/Tensor/Test_Tensor.cpp
+++ b/tests/Unit/DataStructures/Tensor/Test_Tensor.cpp
@@ -1150,7 +1150,7 @@ SPECTRE_TEST_CASE("Unit.DataStructures.Tensor.Multiplicity",
       tensor;
 
   // Test multiplicity by iterator
-  for (auto it = tensor.begin(); it != tensor.end(); ++it) {
+  for (auto it = tensor.cbegin(); it != tensor.cend(); ++it) {
     const auto& indices = tensor.get_tensor_index(it);
     const size_t multiplicity = (indices[1] == indices[2] ? 1 : 2);
     CHECK(multiplicity == tensor.multiplicity(it));
```

However, I don't think there is any guarantee that the Tensor iterators are pointers, so passing `&tensor_element` is not guaranteed to work.  You should explicitly use the iterators rather than a range-based loop.  (You will still need the patch above.)

prayush replied:
> The `multiplicity` function takes an iterator by const-reference.

Yes `multiplicity` takes as input a `const reference` to an `iterator`. The `iterator` itself is pointer without a `const` qualifier. So, gcc deduces that `multiplicity` should take a `const reference` to a `non-const pointer`.  However, `&tensor_element` results in a `const pointer`, and the line above would try to pass to `multiplicity` a `const reference` to that `const pointer` (if i hadn't casted away the `const`). At least, this is what I understood from the error message i get when i don't `const_cast` in the line above:

```
/home/prayush/src/spectre/spectre_fix_rnd_gh_failures/src/DataStructures/Tensor/Tensor.hpp:355:42: note: candidate: constexpr size_t Tensor<X, Symm, IndexList<Indices ...> >::multiplicity(const iterator&) const [with X = DataVector; Symm = brigand::list<>; IndexList = brigand::list; Indices = {}; size_t = long unsigned int; Tensor<X, Symm, IndexList<Indices ...> >::iterator = DataVector*] <near match>
   SPECTRE_ALWAYS_INLINE constexpr size_t multiplicity(
                                          ^~~~~~~~~~~~
/home/prayush/src/spectre/spectre_fix_rnd_gh_failures/src/DataStructures/Tensor/Tensor.hpp:355:42: note:   conversion of argument 1 would be ill-formed:
In file included from /home/prayush/src/spectre/spectre_fix_rnd_gh_failures/tests/Unit/DataStructures/Tensor/EagerMath/Test_Magnitude.cpp:14:0:
/home/prayush/src/spectre/spectre_fix_rnd_gh_failures/src/DataStructures/Tensor/EagerMath/Magnitude.hpp:68:48: **error: invalid conversion from ‘const DataVector*’ to ‘Tensor<DataVector, brigand::list<>, brigand::list<> >::iterator {aka DataVector*}’ [-fpermissive]**
         tensor.multiplicity((&tensor_element)) *  // NOLINT
```


> I don't know if what you are doing here and the `NOLINT` is the right thing to do. Perhaps someone from @sxs-collaboration/spectre-core-devs can have a quick look?

This is one possible workaround. If there's a cleaner way to do this, i'd be happy to learn.

prayush replied:
Aha @wthrowe -- saw your message just now! Thanks for changing `multiplicity`.

prayush replied:
Applied will's patch to make `multiplicity` work with `const iterator`s, and removed the `const_cast` now.

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -141,4 +188,33 @@ struct Sqrt : db::ComputeTag {
+///
+/// \snippet Test_Magnitude.cpp l2norm_name
+template <typename Tag>
+struct L2Norm : db::ComputeTag {
```
Please name this `L2NormCompute` and derive it from an `L2Norm` simple tag. We have been following this pattern for compute tags for a while now, and it seems quite useful so far.

prayush replied:
split

nilsvu commented on tests/Unit/DataStructures/Tensor/EagerMath/Test_Magnitude.cpp:
```diff
@@ -261,6 +267,87 @@ void test_root_tags() {
+  CHECK(Tags::PointwiseL2Norm<Tag>::name() ==
+        "PointwiseL2Norm(" + db::tag_name<Tag>() + ")");
+  CHECK(Tags::L2Norm<Tag>::name() == "L2Norm(" + db::tag_name<Tag>() + ")");
+  /// [l2norm_name]
```
I don't quite understand what this snippet is useful for, also in the other tests in this file. Is the idea to print the `name` to the documentation? @sxs-collaboration/spectre-core-devs are we doing this now?

prayush replied:
Such tests of `SimpleTag`s is quite generic practice in spectre. See for e.g. [here](https://github.com/sxs-collaboration/spectre/blob/develop/tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeSpacetimeQuantities.cpp#L159) or [here](https://github.com/sxs-collaboration/spectre/blob/develop/tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeGhQuantities.cpp#L593) or ...

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -48,11 +50,56 @@ Scalar<DataType> magnitude(
 }
 
+/// \ingroup TensorGroup
+/// \brief Compute point-wise $\\ell^2$ (Frobenius) norm of a tensor
```
Would these functions be better placed in a `Norms.hpp` file? Would do you think @prayush and @fmahebert ?

fmahebert replied:
I agree that placing these functions in  a new `Norms.hpp` would seem more appropriate

prayush replied:
moved code to `Norms.hpp`!

prayush committed:
bugfix in Tensor multiplicity routine

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -48,11 +50,51 @@ Scalar<DataType> magnitude(
+Scalar<DataType> pointwise_l2_norm_square(
+    const Tensor<DataType, Symm, IndexList>& tensor) noexcept {
+  auto pointwise_l2_normsq = make_with_value<DataType>(tensor, 0.);
+  for (auto it = tensor.begin(); it != tensor.end(); ++it) {
```
`it` -> `tensor_element`

prayush replied:
changed

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -48,11 +50,51 @@ Scalar<DataType> magnitude(
+    // stored component with its multiplicity
+    pointwise_l2_normsq += tensor.multiplicity(it) * square(*it);
+  }
+  return Scalar<DataType>{pointwise_l2_normsq};
```
I think this copies the DataVector, so better `std::move(pointwise_l2_normsq)` or `make_with_value<Scalar<DataType>>` to begin with?

prayush replied:
done

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -48,11 +50,51 @@ Scalar<DataType> magnitude(
+/// \brief Compute point-wise $\\ell^2$ (Frobenius) norm of a tensor
+///
+/// \details
+/// Computes the point-wise Frobenius norm of a given tensor of arbitrary rank.
```
Please give a latex expression here to make clear what is computed

fmahebert replied:
I agree: right now the "detail" is the same information as the brief. Please remove this repetition and instead give the mathematical formula being computed.

prayush replied:
Improved documentation for all functions by including formulae, and for compute tags by pointing to those functions.

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -48,11 +50,51 @@ Scalar<DataType> magnitude(
+///
+/// \details
+/// Computes the Frobenius norm of a given tensor of arbitrary rank,
+/// averaged over all grid points.
```
Also give a latex expression here. And move the `averaged over all grid points` to the `\brief`.

fmahebert replied:
Another thing that should be said in this documentation (perhaps in a `\note` or even a `\warning`) is that this is _not_ the L2 norm reduced over the domain, but is the norm on an element.

prayush replied:
Agreed, I've added a `\warning` pointing this out, both in the documentation for `l2_norm` and the associated compute tag

nilsvu commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -141,4 +183,48 @@ struct Sqrt : db::ComputeTag {
+///
+/// \snippet Test_Magnitude.cpp pointwisel2norm_name
+template <typename Tag>
+struct PointwiseL2Norm : db::SimpleTag {
```
Also extend the tags' documentation as you do for the functions

prayush replied:
done

nilsvu reviewed (changes requested):
Please squash the fixups once @fmahebert gives his OK.

fmahebert commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -141,4 +183,48 @@ struct Sqrt : db::ComputeTag {
+/// \ingroup DataStructuresGroup
+/// The point-wise L2 norm of a tensor
+///
+/// \snippet Test_Magnitude.cpp pointwisel2norm_name
```
I'm not sure I understand the goal of this snippet... I think the documentation would be better enhanced with an example of how to observe the norm of a tensor, rather than an example test of the name. Do I misunderstand what you are trying to achieve?

prayush replied:
I agree that this snippet is not very useful, and have removed it

fmahebert reviewed (commented):
I am happy for you to squash previous fixups as well as these changes. (But please keep the bugfix in Tensor as a standalone commit)

I remain unsure about the intended use of this code. Is this primarily for debugging? Or is this intended to live alongside the reduction observers infrastructure? Or is this code eventually going to be called by the "whole domain" norm? Can you clarify any of these questions (here in the Github discussion for now)?

prayush commented:
Hi @nilsleiffischer and @fmahebert - I've addressed your review comments so far. I've also squashed the updates and rebased with develop. This PR is ready for another (hopefully final) look.

prayush commented:
> I remain unsure about the intended use of this code. Is this primarily for debugging? Or is this intended to live alongside the reduction observers infrastructure?

@fmahebert - The intended use of this code is to be able to observe norms of various constraints in volume during an evolution using the generic observing infrastructure provided through `ObserveFields`. Its immediate use will be in #1500 where currently we have a custom observer written that essentially does the same, and duplicates code in the process. Once this PR is merged, that custom observer can be replaced with the generic ones.

> Or is this code eventually going to be called by the "whole domain" norm?

Eventually, probably yes. That would depend on future use cases of this code. For my current purposes (which can be called "debugging", but could be more than that; for instance if one is comparing the evolution of constraint under two different damping formulations, where each produces a "stable" evolution, but one is better than the other).

nilsvu commented on src/DataStructures/Tensor/EagerMath/Norms.hpp:
```diff
@@ -0,0 +1,141 @@
+ * \brief Compute point-wise Frobenius norm of arbitrary tensors.
+ *
+ * \details
+ * At each point \f$ p\f$ in the element, this function computes the
```
Remove extra space before `p`

prayush replied:
done

nilsvu commented on src/DataStructures/Tensor/EagerMath/Norms.hpp:
```diff
@@ -0,0 +1,141 @@
+ * At each point \f$ p\f$ in the element, this function computes the
+ * point-wise Frobenius norm of a given tensor of arbitrary rank. If the tensor
+ * \f$A\f$ has rank \f$n\f$ and dimensionality \f$D\f$, then its Frobenius norm
+ * at point \f$ p\f$ is computed as: \f{equation}
```
Put the `equation` in a new line please. I think it needs a blank line between the paragraph and the `equation` so that clang-format won’t try to shift it back up.

Also remove space before `p`

prayush replied:
done

nilsvu commented on src/DataStructures/Tensor/EagerMath/Norms.hpp:
```diff
@@ -0,0 +1,141 @@
+namespace Tags {
+/*!
+ * \ingroup DataBoxTagsGroup
+ * \ingroup DataStructuresGroup
```
I don’t think any of these tags should be in DataStructuresGroup

prayush replied:
Point-operators on Tensor data structures such as `Tags::Magnitude`, `Tags::Normalized` and `Tags::Sqrt` are presently members of the [data structures doxygen group](https://spectre-code.org/group__datastructuresgroup). `Tags::PointwiseL2Norm` is a similar point-operation on tensors. I can see why `Tags::L2Norm` need not be in the group for data structures, but perhaps `Tags::PointwiseL2Norm` should still live there?

nilsvu replied:
Right, ok seems like this group is indeed the place for these tensor things.. not sure if anyone actually uses it though :) anyway, keep it as is.

nilsvu commented on src/DataStructures/Tensor/EagerMath/Norms.hpp:
```diff
@@ -0,0 +1,141 @@
+/*!
+ * \ingroup DataBoxTagsGroup
+ * \ingroup DataStructuresGroup
+ * Point-wise Frobenius norm of a tensor. See pointwise_l2_norm() for details.
```
Use Doxygens `\see` and put the code in backticks

prayush replied:
done

nilsvu commented on tests/Unit/DataStructures/Tensor/EagerMath/Test_Norms.cpp:
```diff
@@ -0,0 +1,204 @@
+  CHECK(get(db::get<Tags::PointwiseL2Norm<Vector<1, Frame>>>(box)) ==
+        DataVector{npts, 3.});
+  CHECK(get(db::get<Tags::PointwiseL2Norm<Vector<2, Frame>>>(box)) ==
+        DataVector{npts, 4.242640687119285});
```
These should probably be `CHECK_ITERABLE_APPROX` to avoid floating point precision errors.

prayush replied:
changed

nilsvu commented on tests/Unit/DataStructures/Tensor/EagerMath/Test_Norms.cpp:
```diff
@@ -0,0 +1,204 @@
+SPECTRE_TEST_CASE("Unit.DataStructures.Tensor.EagerMath.Norms",
+                  "[DataStructures][Unit]") {
+  test_l2_norm_tag<Frame::Grid>();
+  test_l2_norm_tag<Frame::Inertial>();
```
Does the `Frame` actually change anything?

prayush replied:
This repeats the test for tensors in another frame, and the result do / should not change when compared against hard-coded numbers, which is the point. Testing for 2 frames shows that the L2-norm computation is agnostic to frames.

nilsvu replied:
Does it really test that it's agnostic to frames? For that you would have to take numbers in one frame and compute norms, then coordinate-transform those values and compute norms with the transformed values. And if I'm not mistaken that test is not expected to pass. Or perhaps I misunderstand what you mean with "agnostic to frames".

fmahebert replied:
I think this is testing that the same numerical inputs produce the same numerical outputs, regardless of the Frame parameter. In this sense (!= the physics sense), the norm is independent of Frame. Because the norm doesn't know about the metric, it should give a different answer if the tensors are transformed into a different Frame, where their components will have different numerical values.

nilsvu replied:
Right. Since nothing actually uses the frame parameter I'm not sure if this adds anything to the test.

prayush replied:
Exactly as @fmahebert said, this bit only tests that the same numerical input gives the same numerical output, irrespective of the frame parameter. In other words, it confirms that "nothing actually uses the frame parameter".  That is all it adds to the tests.

nilsvu commented on src/DataStructures/Tensor/EagerMath/Norms.hpp:
```diff
@@ -0,0 +1,141 @@
+ * for details.
+ */
+template <typename Tag>
+struct PointwiseL2NormCompute : PointwiseL2Norm<Tag>, db::ComputeTag {
```
Since we‘ll be computing norms of sets of tensors, I think a version of these tags that works on `Variables` would be useful. This depends on how you‘ll use it Prayush. If you add a whole list of these compute tags to the DataBox where you could add one for a Variables, then please add a Variables overload for this tag. That could also be optimized (perhaps later) to work with the contiguous memory layout of the Variables.

 E.g. `Tags::Flux` in Prefixes.hpp gives an example how to write the Tensor/Variables overload.

prayush replied:
okay, I've added an overload for Variables. (For my use case of observing constraint norms, I *won't* have all constraints available inside a Variables container as they'll each be individually added to the databox by their respective compute tags.)

nilsvu commented on src/DataStructures/Tensor/EagerMath/Norms.hpp:
```diff
@@ -50,23 +59,41 @@ Scalar<DataType> pointwise_l2_norm(
+template <typename TagsList>
+std::array<Scalar<DataVector>, tmpl::size<TagsList>::value> pointwise_l2_norm(
+    const Variables<TagsList>& tensors) noexcept {
+  std::array<Scalar<DataVector>, tmpl::size<TagsList>::value> pointwise_norms;
```
Return a `Variables<db::wrap_tags_in<Tags::PointwiseL2Norm, TagsList>>` here.

nilsvu commented on src/DataStructures/Tensor/EagerMath/Norms.hpp:
```diff
@@ -81,25 +108,55 @@ double l2_norm(const Tensor<DataType, Symm, IndexList>& tensor) noexcept {
               tensor.begin()->size());
 }
+template <typename TagsList>
+std::array<double, tmpl::size<TagsList>::value> l2_norm(
```
Return a `tuples::tagged_tuple_from_typelist<db::wrap_tags_in<L2Norm, TagsList>>` here.

nilsvu commented on src/DataStructures/Tensor/EagerMath/Norms.hpp:
```diff
@@ -81,25 +108,55 @@ double l2_norm(const Tensor<DataType, Symm, IndexList>& tensor) noexcept {
+struct PointwiseL2Norm<Tag,
+                       Requires<tt::is_a_v<::Variables, db::item_type<Tag>>>>
+    : db::SimpleTag {
+  using type = std::array<Scalar<DataVector>,
```
Just `using type = db::item_type<Tag>` (see e.g. `Tags::Flux` in Prefixes.hpp)

nilsvu commented on src/DataStructures/Tensor/EagerMath/Norms.hpp:
```diff
@@ -112,21 +169,36 @@ struct PointwiseL2NormCompute : PointwiseL2Norm<Tag>, db::ComputeTag {
+template <typename Tag>
+struct L2Norm<Tag, Requires<tt::is_a_v<::Variables, db::item_type<Tag>>>>
+    : db::SimpleTag {
+  using type = std::array<double, tmpl::size<typename Tag::tags_list>::value>;
```
also make this a `tuples::TaggedTuple`

nilsvu reviewed (changes requested):
Please squash everything, including these return type changes.

nilsdeppe reviewed (changes requested):
Please remove all of the ones that operator on not `Tensor` or move them elsewhere. `Tensor/EagerMath` should be exactly that: operations done on tensors.

nilsvu commented:
@nilsdeppe do you mean the `Variables` overloads? These overloads should probably stick together.. Do you have a suggestion where to move things?

nilsdeppe commented:
Yes, no, and I don't know why you even have a Variables version. What's the use case?

nilsvu commented:
We observe norms of all system variables errors (to the analytic solution), for instance, so I think it's nice to have a compute tag for the Variables instead of adding individual compute tags for everything.

nilsdeppe commented:
Yea, let's not have that. Please just remove the variables overloads.

nilsvu commented:
Ok. Please squash.

fmahebert commented on src/DataStructures/Tensor/EagerMath/Norms.hpp:
```diff
@@ -0,0 +1,213 @@
+ * At each point \f$p\f$ in the element, this function computes the
+ * point-wise Frobenius norm(s) of a given Tensor with arbitrary rank (or a
+ * Variables of Tensors). If the Tensor \f$A\f$ has rank \f$n\f$ and
+ * dimensionality \f$D\f$, then its Frobenius norm at point \f$ p\f$ is
```
Remove the extra space before the `p`

prayush replied:
removed

fmahebert commented on src/DataStructures/Tensor/EagerMath/Norms.hpp:
```diff
@@ -0,0 +1,213 @@
+ * computed as:
+ *
+ * \f{equation}
+ * ||A||_2^p =
```
I find this notation surprising. Do we denote grid points by superscripts anywhere else? I strongly suggest changing `p` to a subscript.

prayush replied:
I had superscripted (instead of subscripting) `p` to disambiguate it from tensorial indices in subscript. I've now moved `p` from superscript to argument notation `(p)`, which seems right in the context of these equations.

fmahebert commented on tests/Unit/DataStructures/Tensor/EagerMath/Test_Norms.cpp:
```diff
@@ -0,0 +1,311 @@
+
+  // Check names
+  using Tag = MyScalar;
+  /// [l2norm_name]
```
I think this tag is now unused, so you could remove it (also the one below)

prayush replied:
done

fmahebert commented on src/DataStructures/Tensor/EagerMath/Norms.hpp:
```diff
@@ -0,0 +1,213 @@
+// @{
+/*!
+ * \ingroup TensorGroup
+ * \brief Compute point-wise Frobenius norm of arbitrary Tensors.
```
I have a mild preference for changing the brief to the more accessible "Compute point-wise Euclidean L2 norm of arbitrary Tensors". I do not feel strongly about this suggestion, so let me know if you would rather keep as it is. I am happy say "Frobenius" in the details below, as is currently worded.

prayush replied:
done

fmahebert commented on src/DataStructures/Tensor/EagerMath/Norms.hpp:
```diff
@@ -0,0 +1,213 @@
+// @{
+/*!
+ * \ingroup TensorGroup
+ * \brief Compute Frobenius norm of arbitrary Tensors reduced over an element.
```
Some note re "Frobenius"

prayush replied:
done

fmahebert reviewed (commented):
I'm also happy for you to squash, including these documentation notes.

prayush commented:
@nilsdeppe @nilsleiffischer @fmahebert - I've removed the Variables overload, and squashed all changes.

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/Norms.hpp:
```diff
@@ -0,0 +1,150 @@
+ * \details
+ * Computes the RMS value of the point-wise Frobenius norm of a given Tensor
+ * with arbitrary rank over all grid points in an element. If the Tensor \f$A\f$
+ * has rank \f$n\f$ and dimensionality \f$D\f$, and the element has \f$N\f$
```
were using `N` as the order of the element, so the points would be `N+1`. While you do state what it means, I think it's better to avoid mixing these too since there was a lot of confusion for a long time about expected order of convergence because some papers use N for the points and others N for the order. Maybe `N_p` for the points? Thoughts?

prayush replied:
this is a fair point.  I can use `N_p` for number of points or just use `N` as the summation's upper limit instead of `N-1` and state clearly that `... and the element (of order \f$N\f$) has \f$N+1\f$ points...`. Let me know which you think would be more coherent with the rest of the code. I mildly prefer the latter since it refrains from introducing yet another variable `N_p` when `N+1` should suffice.

nilsdeppe replied:
I agree, using `N+1` seems better, but I'm okay with either :)

prayush replied:
changed

nilsdeppe commented on tests/Unit/DataStructures/Tensor/EagerMath/Test_Norms.cpp:
```diff
@@ -0,0 +1,219 @@
+void test_l2_norm_tag() {
+  constexpr size_t npts = 5;
+
+  const DataVector one(npts, 1.);
```
Maybe I'm not seeing this test, but I'm not convinced (because I don't see a test for it) that if I have a `DataVector{1.0, 2.0, 3.0}` that I will get the right now. That is, the pointwise norm should not be tested on only uniform `DataVector`s. I don't think this needs to be done extensively, one test on a vector (in 1,2, and 3d) would be enough for me :)

prayush replied:
added

nilsdeppe replied:
Thanks!

nilsdeppe reviewed (commented):
One minor dox comment and I'm requesting an additional test (unless I overlooked that it already exists). I'm happy for you to squash :) Overall, looks good :)

nilsdeppe reviewed (approved):
LGTM!
